### PR TITLE
feat(2319): [2] Support template composition. BREAKING CHANGE: must pass in template factory

### PR DESCRIPTION
**Pull request by tkyi**
## Context

It would be nice if template owners could use templates in their template configs.

## Objective

This PR fetches template if it exists in the job config section and merges it accordingly.
Also merged `images` field.

## References

Related to https://github.com/screwdriver-cd/screwdriver/issues/2319
Blocked by https://github.com/screwdriver-cd/data-schema/pull/431

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: lib/helper.js [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+async function mergeTemplateIntoJob(templateObj, templateFactory) {
+    // Child template config
+    const oldJob = templateObj.config;
+
+    // Try to get the template
+    return templateFactory.getTemplate(oldJob.template)
+        .then((template) => {
+            if (!template) {
+                throw new Error(`Template ${oldJob.template} does not exist`);
+            }
```
Comment (klu909) [inline lib/helper.js:181]: here we can also check child template's `order` against locked steps before merging two templates

File: lib/helper.js [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    newJob.environment = newJob.environment || {};
+    newJob.settings = newJob.settings || {};
+
+    // Merge
+    Object.assign(newJob.annotations, oldJob.annotations || {});
+    Object.assign(newJob.environment, oldJob.environment || {});
+    Object.assign(newJob.settings, oldJob.settings || {});
+    newJob.image = oldJob.image || newJob.image;
+
+    if (oldJob.requires) {
```
Comment (klu909) [inline lib/helper.js:41]: are line 41 to line 55 necessary for merging templates?

File: lib/helper.js [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            if (key.startsWith('teardown-')) {
+                teardownSteps.push(key);
+            }
+
+            obj[key] = item[key];
+
+            return obj;
+        }, {});
+
+        for (let i = 0; i < newJob.steps.length; i += 1) {
```
Comment (klu909) [inline lib/helper.js:127]: I don't think child template should have pre/post, if necessary people should just use `order` field.

File: lib/helper.js (hunk 1 above)
Comment (tkyi) [reply lib/helper.js:181]: I think it'll be better to add locked steps feature in a separate PR

File: lib/helper.js (hunk 2 above)
Comment (tkyi) [reply lib/helper.js:41]: these are all fields that could exist in a template, right?

File: lib/helper.js (hunk 3 above)
Comment (tkyi) [reply lib/helper.js:127]: I'm okay with your proposal. What do you think about `teardown-` steps?

@jithine what do you think?

File: lib/helper.js (hunk 2 above)
Comment (klu909) [reply lib/helper.js:41]: cache might make sense, but for others I don't think anyone would expect them in templates.

File: lib/helper.js (hunk 3 above)
Comment (jithine) [reply lib/helper.js:127]: order should be optional.
1.  If order is not used, then regular step merging should happen.
2. If order is used, only order should be used to determine step order, no need for pre/post
3. teardown- should continue to work as it is currently. at the end of all user steps.

File: lib/helper.js (hunk 2 above)
Comment (tkyi) [reply lib/helper.js:41]: I think it's okay to leave it
if template owners choose to use it, template users just need to be aware

File: lib/helper.js (hunk 3 above)
Comment (tkyi) [reply lib/helper.js:127]: 1. This is already in my PR
2. This is already in my PR as well, though I didn't explicitly prevent pre/post steps
3. Added support for teardown-

**Review by klu909 — APPROVED** [review]

**Pull request merged**